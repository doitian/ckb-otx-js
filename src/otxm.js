/** @module otxm */
import { molecule, number } from "@ckb-lumos/codec";
import { BI } from "@ckb-lumos/bi";
import { blockchain } from "@ckb-lumos/base";

const { table, vector } = molecule;
export const { BytesOpt, Bytes, Byte32, Script, ScriptOpt, OutPoint } =
  blockchain;
export const { Uint32LE, Uint64LE, Uint128LE } = number;

/**
 * Numbers that can be converted into BI, examples:
 * - Number in hex string: `"0x1"`
 * - JavaScript Number: `1`
 * - Number in decimal string: `"1"`
 * @typedef {import("@ckb-lumos/bi").BIish} BIish
 */

/**
 * Bytes like data, examples:
 * - Hex string: `"0x68656c6c6f"`
 * - Uint8Array: `[0x68, 0x65, 0x6c, 0x6c, 0x6f]`
 * @typedef {import("@ckb-lumos/codec").BytesLike} BytesLike
 */

/**
 * @typedef {Object} OtxKeyPairPOJO
 * @property {BIish} keyType
 * @property {BytesLike=} keyData
 * @property {BytesLike} valueData
 */

/**
 * @callback packOtxKeyPair
 * @param {OtxKeyPairPOJO} pojo
 * @returns {Uint8Array}
 */

/**
 * @typedef {Object} OtxKeyPair
 * @property {function(OtxKeyPairPOJO): Uint8Array} pack `(pojo: [OtxKeyPairPOJO]{@link module:otxm~OtxKeyPairPOJO}) => Uint8Array` packs OtxKeyPairPojo into molecule buffer.
 * @property {function(BytesLike): OtxKeyPairPOJO} unpack `(buf: [BytesLike]{@link module:otxm~BytesLike}) => [OtxKeyPairPOJO]{@link module:otxm~OtxKeyPairPOJO}` unpacks molecule buffer into OtxKeyPairPOJO.
 */

/**
 * Molecule packer for tye type OtxKeyPair.
 *
 * ```
 * table OtxKeyPair {
 *   keyType: Uint32,
 *   keyData: BytesOpt,
 *   valueData: Bytes,
 * }
 * ```
 *
 * @type {OtxKeyPair}
 */
export const OtxKeyPair = table(
  {
    keyType: Uint32LE,
    keyData: BytesOpt,
    valueData: Bytes,
  },
  ["keyType", "keyData", "valueData"]
);

/**
 * @typedef {Object} OtxMap
 * @property {function(OtxKeyPairPOJO[]): Uint8Array} pack `(pojo: [OtxKeyPairPOJO]{@link module:otxm~OtxKeyPairPOJO}[]) => Uint8Array` packs OtxKeyPairPojo[] into molecule buffer.
 * @property {function(BytesLike): OtxKeyPairPOJO[]} unpack `(buf: [BytesLike]{@link module:otxm~BytesLike}) => [OtxKeyPairPOJO]{@link module:otxm~OtxKeyPairPOJO}[]` unpacks molecule buffer into OtxKeyPairPOJO[].
 */

/**
 * Molecule packer for tye type OtxMap.
 *
 * ```
 * vector OtxMap <OtxKeyPair>;
 * ```
 *
 * @type {OtxMap}
 */
export const OtxMap = vector(OtxKeyPair);

/**
 * @typedef {Object} OtxMapVec
 * @property {function(OtxKeyPairPOJO[][]): Uint8Array} pack `(pojo: [OtxKeyPairPOJO]{@link module:otxm~OtxKeyPairPOJO}[][]) => Uint8Array` packs OtxKeyPairPojo[][] into molecule buffer.
 * @property {function(BytesLike): OtxKeyPairPOJO[][]} unpack `(buf: [BytesLike]{@link module:otxm~BytesLike}) => [OtxKeyPairPOJO]{@link module:otxm~OtxKeyPairPOJO}[][]` unpacks molecule buffer into OtxKeyPairPOJO[][].
 */

/**
 * Molecule packer for tye type OtxMap.
 *
 * ```
 * vector OtxMapVec <OtxMap>;
 * ```
 *
 * @type {OtxMapVec}
 */
export const OtxMapVec = vector(OtxMap);

/**
 * @typedef {Object} OpenTransactionPOJO
 * @property {OtxKeyPairPOJO[]} meta
 * @property {OtxKeyPairPOJO[][]} cellDeps
 * @property {OtxKeyPairPOJO[][]} headerDeps
 * @property {OtxKeyPairPOJO[][]} inputs
 * @property {OtxKeyPairPOJO[][]} witnesses
 * @property {OtxKeyPairPOJO[][]} outputs
 */

/**
 * @typedef {Object} OpenTransaction
 * @property {function(OpenTransactionPOJO): Uint8Array} pack `(pojo: [OpenTransactionPOJO]{@link module:otxm~OpenTransactionPOJO}) => Uint8Array` packs OpenTransactionPOJO into molecule buffer.
 * @property {function(BytesLike): OpenTransactionPOJO} unpack `(buf: [BytesLike]{@link module:otxm~BytesLike}) => [OpenTransactionPOJO]{@link module:otxm~OpenTransactionPOJO}` unpacks molecule buffer into OpenTransactionPOJO.
 */

/**
 * Molecule packer for tye type OpenTransaction.
 *
 * ```
 * table OpenTransaction {
 *   meta: OtxMap,
 *   cell_deps: OtxMapVec,
 *   header_deps: OtxMapVec,
 *   inputs: OtxMapVec,
 *   witnesses: OtxMapVec,
 *   outputs: OtxMapVec,
 * }
 * ```
 *
 * @type {OpenTransaction}
 */
export const OpenTransaction = table(
  {
    meta: OtxMap,
    cellDeps: OtxMapVec,
    headerDeps: OtxMapVec,
    inputs: OtxMapVec,
    witnesses: OtxMapVec,
    outputs: OtxMapVec,
  },
  ["meta", "cellDeps", "headerDeps", "inputs", "witnesses", "outputs"]
);
