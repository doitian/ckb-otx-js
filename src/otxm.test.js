import { OtxKeyPair, OtxMap, OtxMapVec, OpenTransaction } from "./otxm.js";

test("OtxKeyPair without keyData", () => {
  const pojo = { keyType: 1, valueData: "0x0f" };
  const buf = OtxKeyPair.pack(pojo);
  expect(buf).toMatchSnapshot();
  expect(OtxKeyPair.unpack(buf)).toEqual(pojo);
});

test("OtxKeyPair with keyData", () => {
  const pojo = { keyType: 1, keyData: "0x0a", valueData: "0x0f" };
  const buf = OtxKeyPair.pack(pojo);
  expect(buf).toMatchSnapshot();
  expect(OtxKeyPair.unpack(buf)).toEqual(pojo);
});

test("OtxMap", () => {
  const pojo = [
    { keyType: 1, keyData: "0x0a", valueData: "0x0e" },
    { keyType: 2, valueData: "0x0f" },
  ];
  const buf = OtxMap.pack(pojo);
  expect(buf).toMatchSnapshot();
  expect(OtxMap.unpack(buf)).toEqual(pojo);
});

test("OtxMapVec", () => {
  const pojo = [
    [
      { keyType: 1, keyData: "0x0a", valueData: "0x0e" },
      { keyType: 2, valueData: "0x0f" },
    ],
    [
      { keyType: 3, keyData: "0xfa", valueData: "0xfe" },
      { keyType: 4, valueData: "0xff" },
    ],
  ];
  const buf = OtxMapVec.pack(pojo);
  expect(buf).toMatchSnapshot();
  expect(OtxMapVec.unpack(buf)).toEqual(pojo);
});

test("OpenTransaction", () => {
  const pojo = {
    meta: [
      { keyType: 1, keyData: "0x0a", valueData: "0x0e" },
      { keyType: 2, valueData: "0x0f" },
    ],
    cellDeps: [[{ keyType: 3, valueData: "0xff03" }]],
    headerDeps: [[{ keyType: 4, valueData: "0xff04" }]],
    inputs: [[{ keyType: 5, valueData: "0xff05" }]],
    witnesses: [[{ keyType: 6, keyData: "0x06", valueData: "0xff06" }]],
    outputs: [[{ keyType: 7, keyData: "0x07", valueData: "0xff07" }]],
  };

  const buf = OpenTransaction.pack(pojo);
  expect(buf).toMatchSnapshot();
  expect(OpenTransaction.unpack(buf)).toEqual(pojo);
});
