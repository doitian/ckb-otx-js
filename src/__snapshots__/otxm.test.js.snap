// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OpenTransaction 1`] = `
Uint8Array [
  59,
  1,
  0,
  0,
  28,
  0,
  0,
  0,
  95,
  0,
  0,
  0,
  137,
  0,
  0,
  0,
  179,
  0,
  0,
  0,
  221,
  0,
  0,
  0,
  12,
  1,
  0,
  0,
  67,
  0,
  0,
  0,
  12,
  0,
  0,
  0,
  42,
  0,
  0,
  0,
  30,
  0,
  0,
  0,
  16,
  0,
  0,
  0,
  20,
  0,
  0,
  0,
  25,
  0,
  0,
  0,
  1,
  0,
  0,
  0,
  1,
  0,
  0,
  0,
  10,
  1,
  0,
  0,
  0,
  14,
  25,
  0,
  0,
  0,
  16,
  0,
  0,
  0,
  20,
  0,
  0,
  0,
  20,
  0,
  0,
  0,
  2,
  0,
  0,
  0,
  1,
  0,
  0,
  0,
  15,
  42,
  0,
  0,
  0,
  8,
  0,
  0,
  0,
  34,
  0,
  0,
  0,
  8,
  0,
  0,
  0,
  26,
  0,
  0,
  0,
  16,
  0,
  0,
  0,
  20,
  0,
  0,
  0,
  20,
  0,
  0,
  0,
  3,
  0,
  0,
  0,
  2,
  0,
  0,
  0,
  255,
  3,
  42,
  0,
  0,
  0,
  8,
  0,
  0,
  0,
  34,
  0,
  0,
  0,
  8,
  0,
  0,
  0,
  26,
  0,
  0,
  0,
  16,
  0,
  0,
  0,
  20,
  0,
  0,
  0,
  20,
  0,
  0,
  0,
  4,
  0,
  0,
  0,
  2,
  0,
  0,
  0,
  255,
  4,
  42,
  0,
  0,
  0,
  8,
  0,
  0,
  0,
  34,
  0,
  0,
  0,
  8,
  0,
  0,
  0,
  26,
  0,
  0,
  0,
  16,
  0,
  0,
  0,
  20,
  0,
  0,
  0,
  20,
  0,
  0,
  0,
  5,
  0,
  0,
  0,
  2,
  0,
  0,
  0,
  255,
  5,
  47,
  0,
  0,
  0,
  8,
  0,
  0,
  0,
  39,
  0,
  0,
  0,
  8,
  0,
  0,
  0,
  31,
  0,
  0,
  0,
  16,
  0,
  0,
  0,
  20,
  0,
  0,
  0,
  25,
  0,
  0,
  0,
  6,
  0,
  0,
  0,
  1,
  0,
  0,
  0,
  6,
  2,
  0,
  0,
  0,
  255,
  6,
  47,
  0,
  0,
  0,
  8,
  0,
  0,
  0,
  39,
  0,
  0,
  0,
  8,
  0,
  0,
  0,
  31,
  0,
  0,
  0,
  16,
  0,
  0,
  0,
  20,
  0,
  0,
  0,
  25,
  0,
  0,
  0,
  7,
  0,
  0,
  0,
  1,
  0,
  0,
  0,
  7,
  2,
  0,
  0,
  0,
  255,
  7,
]
`;

exports[`OtxKeyPair with keyData 1`] = `
Uint8Array [
  30,
  0,
  0,
  0,
  16,
  0,
  0,
  0,
  20,
  0,
  0,
  0,
  25,
  0,
  0,
  0,
  1,
  0,
  0,
  0,
  1,
  0,
  0,
  0,
  10,
  1,
  0,
  0,
  0,
  15,
]
`;

exports[`OtxKeyPair without keyData 1`] = `
Uint8Array [
  25,
  0,
  0,
  0,
  16,
  0,
  0,
  0,
  20,
  0,
  0,
  0,
  20,
  0,
  0,
  0,
  1,
  0,
  0,
  0,
  1,
  0,
  0,
  0,
  15,
]
`;

exports[`OtxMap 1`] = `
Uint8Array [
  67,
  0,
  0,
  0,
  12,
  0,
  0,
  0,
  42,
  0,
  0,
  0,
  30,
  0,
  0,
  0,
  16,
  0,
  0,
  0,
  20,
  0,
  0,
  0,
  25,
  0,
  0,
  0,
  1,
  0,
  0,
  0,
  1,
  0,
  0,
  0,
  10,
  1,
  0,
  0,
  0,
  14,
  25,
  0,
  0,
  0,
  16,
  0,
  0,
  0,
  20,
  0,
  0,
  0,
  20,
  0,
  0,
  0,
  2,
  0,
  0,
  0,
  1,
  0,
  0,
  0,
  15,
]
`;

exports[`OtxMapVec 1`] = `
Uint8Array [
  146,
  0,
  0,
  0,
  12,
  0,
  0,
  0,
  79,
  0,
  0,
  0,
  67,
  0,
  0,
  0,
  12,
  0,
  0,
  0,
  42,
  0,
  0,
  0,
  30,
  0,
  0,
  0,
  16,
  0,
  0,
  0,
  20,
  0,
  0,
  0,
  25,
  0,
  0,
  0,
  1,
  0,
  0,
  0,
  1,
  0,
  0,
  0,
  10,
  1,
  0,
  0,
  0,
  14,
  25,
  0,
  0,
  0,
  16,
  0,
  0,
  0,
  20,
  0,
  0,
  0,
  20,
  0,
  0,
  0,
  2,
  0,
  0,
  0,
  1,
  0,
  0,
  0,
  15,
  67,
  0,
  0,
  0,
  12,
  0,
  0,
  0,
  42,
  0,
  0,
  0,
  30,
  0,
  0,
  0,
  16,
  0,
  0,
  0,
  20,
  0,
  0,
  0,
  25,
  0,
  0,
  0,
  3,
  0,
  0,
  0,
  1,
  0,
  0,
  0,
  250,
  1,
  0,
  0,
  0,
  254,
  25,
  0,
  0,
  0,
  16,
  0,
  0,
  0,
  20,
  0,
  0,
  0,
  20,
  0,
  0,
  0,
  4,
  0,
  0,
  0,
  1,
  0,
  0,
  0,
  255,
]
`;
